### ISSUE TITLE
Implement HasAriaLabel interface in components

### ISSUE THREAD

**Issue by web-padawan** [vaadin/flow-components#2946]
### Motivation

Part of https://github.com/vaadin/platform/issues/3803
Part of https://github.com/vaadin/platform/issues/3814

Originally requested at #2160

The `HasAriaLabel` interface has been added to Flow: https://github.com/vaadin/flow/pull/12126

For now we only have relevant API on the `Checkbox` (and it's not behaving correctly, see the list below):

https://github.com/vaadin/flow-components/blob/fed476de4cee4b0f885e13e02f268c454cc33013/vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/Checkbox.java#L167-L169

### Proposed Solution

Update components to implement `HasAriaLabel` where it makes sense. 
Note, the element where to set `aria-label`  depends on the html structure in every field.

#### `aria-label` on the host

The following components would need this attribute to be set on the host:

- `Button`
- `CheckboxGroup`
- `RadioButtonGroup`
- `Tab`
- (any others?)

#### `aria-label` on the focusable child

These components need `aria-label` to be set on the `<input>`, `<textarea>` or `<vaadin-select-value-button>`

- `Checkbox`
- `ComboBox`
- `DatePicker`
- `TextField` + other fields
- `Select`
- `TimePicker`

#### Complex components

These components consist of multiple fields so we need to decide where to set `aria-label`.

- `CustomField`
- `DateTimePicker`

**Comment by DiegoCardoso** [comment]
A few topics to discuss regarding the implementation of this issue:

### `HasAriaLabel` API

The `HasAriaLabel` interface now only has `setAriaLabel(String)` and `getAriaLabel()` on it, but the AC also mentions support for defining the `aria-labelledby` attribute. There are two API proposals to achieve this:
1. Add a method overloading defined as `setAriaLabel(HasElement)`
2. Add a new pair of setter and getter defined as `setAriaLabelledBy(HasElement)` and `getAriaLabelledBy()`

Per [the spec](https://w3c.github.io/aria/#aria-labelledby), if both `aria-label` and `aria-labelledby`, the latter takes precedence over the former:
> As required by the [accessible name and description computation](https://w3c.github.io/aria/#textalternativecomputation), user agents give precedence to [aria-labelledby](https://w3c.github.io/aria/#aria-labelledby) over [aria-label](https://w3c.github.io/aria/#aria-label) when computing the accessible name property.

With that in mind, option (1) can make it a bit clearer that you should use one or the other (`setAriaLabel(String)` vs `setAriaLabel(HasElement)`), but makes it unclear what `getAriaLabel()` should return when `setAriaLabel(HasElement)` is used.

Option (2) can give the impression that both can be used together, which would require a bit more work on the Java documentation of the API.

### `setLabel(String)` vs `setAriaLabel(String)` vs `setAriaLabel(HasElement)` / `setAriaLabelledBy(HasElement)`

Another issue worth discussing is the interoperability of the different methods to define a label for the component. When a label is defined to an input field component (eg. `TextField`), the value is added as the text content of the `<label>` element and then its `id` is set to the `aria-labelledby` attribute of the input element. The same happens when a `<label slot="label">` element is created inside a `<vaadin-form-item>`. Because of the precedence issue described in the previous session, this leads to open questions like:
- If `setLabel(String)` (or `FormItem` with label) is used, what should happen if `setAriaLabel(String)` is called? Should it clear `aria-labelledby` on the input element?
- When `setAriaLabel(HasElement)`/`setAriaLabelledBy(HasElement)` is used, it should replace the `aria-labelledby` with the new element's `id`. What should happen if it's called with `null`? I guess it should restore the previous value, but are there guarantees that the previous value still holds true?

**Comment by rolfsmeds** [comment]
One detail that might be useful regarding the first question, `setAriaLabel(HasElement)` vs `setAriaLabelledBy(HasElement)`, is that the getter would be _very_ rarely used (just as it's quite rare to use the `getLabel()` method for the visible label). With that in mind, I don't see it as a DX issue if `getAriaLabel()` would return an `Object`, which the developer would then have to cast to a `String` or a `HasElement`.

As for the second question, whether the normal `aria-labelledby` association set automatically to the built-in label element (or the `FormItem`'s label) should be removed when `aria-label` is set manually, there is at least one use case in favor of removing the normal `labelledby`: Sometimes a separate `aria-label` is set to an element that already has a visible label to provide more context to screen reader users – context that may be quite obvious e.g. from the layout context for a sighted user, such as an `aria-label="Search open issues"` on a search field above a list of open issues, which has a shorter visible `"Search"` label for sighted users. That use case can only be catered for by removing the normal label from being announced, as it will otherwise replace the manually set `aria-label`.

It would of course be nice if, when the `aria-label` is removed, the normal `aria-labelledby` would be reinstated. (And the same goes for when a custom `aria-labelledby` element is set.)

**Comment by DiegoCardoso** [comment]
> With that in mind, I don't see it as a DX issue if getAriaLabel() would return an Object, which the developer would then have to cast to a String or a HasElement.

Since `HasAriaLabel#getAriaLabel()` returns `Optional<String>`, that would be a breaking change, unless we decide to not use this interface.

**Comment by knoobie** [comment]
I would suggest to stay close to client side and just provide String based get/set methods (added/adding to the HasAriaLabel interface). Supplying methods based on HasElements as param suggests that literally everything could supply an ID, even tho some elements aren't meant for it or users forgot to add an unique ID to them.. or are they expecting Vaadin to handle / create / make sure elements added as aria-labelledBy have a unique ID. This feels dangerous and error prone. 

> Since HasAriaLabel#getAriaLabel() returns Optional<String>

I'm sorry 😬 I would have loved to return String there, but I think Denis had the same idea as Rolf, that most people probably never call the getter..

**Comment by Legioth** [comment]
I would assume that most of those who would use this API through Flow are already quite proficient with how ARIA attributes work on the DOM level. And if they aren't they will probably end up reading generic documentation on that topic. It would be confusing for them if we use different naming and semantics that what's used in the specification. This thinking leads me towards suggesting separate `setAriaLabel` and `setAriaLabelledBy` methods.

I also support the thought of making `setAriaLabelledBy` use `String` rather than `Element` or `HasElement`. This is in part to stay closer to the spec and in part to avoid complexity such whether we should make sure the attribute value is also updated in case `labelComponent.setId("foo")` is called later.

**Comment by rolfsmeds** [comment]
As a frontend-adjacent guy myself, I definintely understand the "staying close to client side / native" sentiment, but I'm also painfully aware that that is what we did with Horizontal and VerticalLayout, and those are not exactly gleaming beacons of DX success.

I was actually hoping to provide an API that would require as little knowledge about ARIA attributes and element `id`s as possible, to encourage more Flow developers to provide ARIA labels – not just the 1% who are have read up on the spec. Unless I'm mistaken, we already do things like track attribute changes in other features.

But, of course, we could provide just `setAriaLabelledBy(String)` for now, and later add a `setAriaLabelledBy(HasElement)` convenience override, if there's indication that faffing around with `id`s is seen as an inconvenience by Flow devs.

So, FWIW, I'm fine with `setAriaLabelledBy(String)` for now – it's already miles better than the javascript hack you need to resort to today – but I'm not convinced that that is the best we could do in the long term.

**Comment by web-padawan** [comment]
All the PRs linked to this issue have landed in Vaadin 24.1 stable which is out today. Closing this issue as fixed.
